src/Nethermind/Nethermind.Serialization.Rlp/Rlp.cs

```diff
@@ -234,10 +232,10 @@ internal static ArrayPoolList<byte> ByteSpanToArrayPool(ReadOnlySpan<byte> span)
             return span.ToPooledList();
         }
 
-        public static IRlpValueDecoder<T>? GetValueDecoder<T>() => Decoders.TryGetValue(typeof(T), out IRlpDecoder value) ? value as IRlpValueDecoder<T> : null;
-        public static IRlpStreamDecoder<T>? GetStreamDecoder<T>() => Decoders.TryGetValue(typeof(T), out IRlpDecoder value) ? value as IRlpStreamDecoder<T> : null;
-        public static IRlpObjectDecoder<T>? GetObjectDecoder<T>() => Decoders.TryGetValue(typeof(T), out IRlpDecoder value) ? value as IRlpObjectDecoder<T> : null;
-        public static IRlpDecoder<T>? GetDecoder<T>() => Decoders.TryGetValue(typeof(T), out IRlpDecoder value) ? value as IRlpDecoder<T> : null;
+        public static IRlpValueDecoder<T>? GetValueDecoder<T>(string key = RlpDecoderKey.Default) => Decoders.TryGetValue(new(typeof(T), key), out IRlpDecoder value) ? value as IRlpValueDecoder<T> : null;
```

Comment:
I see several places where these methods are used to assign non-nullable fields.
How about adding a version that throws if decoder is missing, but never returns `null`? Instead of potential null-reference errors later on?

Reply:
Yeah, it would be cool. But why it's like that in the first place? I'm postponing such further improvements, because we actually need to replace it fully, with real DI like in @asdacap's PRs